### PR TITLE
implement no-mutating-reducer-state rule

### PR DESCRIPTION
**Pull request by rayhanadev**
## Why

Catches reducers that mutate the current state object and then return that same object.

React compares reducer state by reference. If a reducer changes the old object in place and returns it, React can treat the update as unchanged. That can skip rerenders or leave memoized consumers with stale data.

Before, this reducer mutates `state.items` and returns the same `state` object, so the rule reports it:

```tsx
import { useReducer } from "react";

function reducer(state, action) {
  if (action.type === "add") {
    state.items.push(action.item);
    return state;
  }

  return state;
}

useReducer(reducer, { items: [] });
```

After, the reducer returns a new state object, so the rule does not report it:

```tsx
import { useReducer } from "react";

function reducer(state, action) {
  if (action.type === "add") {
    return {
      ...state,
      items: [...state.items, action.item],
    };
  }

  return state;
}

useReducer(reducer, { items: [] });
```

## What changed

- Added `no-mutating-reducer-state` under State & Effects.
- Detects real React `useReducer` calls and same-file reducer functions.
- Tracks aliases to the original state, like `const next = state`.
- Reports common mutations when the same state reference can be returned.
- Allows safe patterns like clone-first updates, no-op `return state`, Immer-style wrappers, and non-React `useReducer` calls.
- Added tests for valid bugs, false-positive guards, and eval-derived edge cases.

## Test plan

```
pnpm exec vp test run packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.test.ts
pnpm --filter oxlint-plugin-react-doctor typecheck
pnpm lint
```

### CODE REVIEW

**Comment by rayhanadev** [comment]
bugbot run

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+    )
+      return;
+    if (isExpressionRootedInMutableReducerStateSource(child.callee.object, state)) {
+      mutations.push({ node: child });
+    }
+  });
+  return mutations;
+};
+
+// Updates alias sets for declarations such as `const alias = state` and
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: no

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+    }
+    if (!isNodeOfType(child.callee, "MemberExpression")) return;
+    const methodName = getStaticMemberPropertyName(child.callee);
+    // Receiver-mutating methods mutate the object/array/collection they are
+    // called on. We only record them when the receiver is state-derived:
+    //
+    //   state.items.push(item);
+    //   items.splice(index, 1);
+    //   stateMap.set(key, value);
+    //
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts:424]: good

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+  let current: EsTreeNode | null | undefined = node;
+  while (current && isNodeOfType(current, "MemberExpression")) {
+    current = current.object;
+  }
+  return (
+    isNodeOfType(current, "Identifier") && state.mutableStateSourceNames.has(current.name)
+  );
+};
+
+// Tracks top-level identity only: identifiers in this set are references React
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: no

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 4]
```diff
@@ -0,0 +1,645 @@
...
+// React.useReducer as reducer functions.
+const isFunctionLikeAstNode = (node: EsTreeNode | null | undefined): node is FunctionLikeNode =>
+  Boolean(
+    node &&
+      (isNodeOfType(node, "FunctionDeclaration") ||
+        isNodeOfType(node, "FunctionExpression") ||
+        isNodeOfType(node, "ArrowFunctionExpression")),
+  );
+
+// Checks whether an import specifier's enclosing import declaration is from
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: no

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 5]
```diff
@@ -0,0 +1,645 @@
...
+  mutableStateSourceNames: Set<string>;
+  mutations: ReducerStateMutation[];
+}
+
+type FunctionLikeNode =
+  | EsTreeNodeOfType<"FunctionDeclaration">
+  | EsTreeNodeOfType<"FunctionExpression">
+  | EsTreeNodeOfType<"ArrowFunctionExpression">;
+
+// Forks path state when control flow branches. Each path carries its own alias
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: no

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 6]
```diff
@@ -0,0 +1,645 @@
...
+
+const REFLECT_MUTATION_METHODS = new Set(["deleteProperty", "set"]);
+
+const OBJECT_ASSIGN_METHOD = new Set(["assign"]);
+
+// React reducer state is compared by identity (`Object.is`). A reducer may
+// legitimately return the previous state object for no-op actions, and it may
+// legitimately mutate freshly-cloned data before returning the clone. The bug
+// this rule targets is narrower:
+//
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts:42]: good

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 7]
```diff
@@ -0,0 +1,645 @@
...
+      ) {
+        mutations.push({ node: child });
+      }
+      return;
+    }
+
+    if (isNodeOfType(child, "UpdateExpression")) {
+      // Updates are writes too:
+      //
+      //   state.count++;
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts:375]: good

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+
+  const reportReducerStateMutations = (mutations: ReducerStateMutation[]): void => {
+    for (const mutation of mutations) {
+      if (reportedNodes.has(mutation.node)) continue;
+      reportedNodes.add(mutation.node);
+      context.report({ node: mutation.node, message: MESSAGE });
+    }
+  };
+
+  // Returns the paths that are still alive after these statements run.
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: no

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 9]
```diff
@@ -0,0 +1,645 @@
...
+    "Return a new reducer state object instead of mutating the current state and returning the same reference. React uses object identity to decide whether reducer state changed.",
+  create: (context: RuleContext) => {
+    const analyzedReducers = new WeakSet<EsTreeNode>();
+    const reportedNodes = new WeakSet<EsTreeNode>();
+
+    return {
+      CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
+        // Pipeline:
+        // 1. accept only calls proven to be React's imported useReducer;
+        // 2. resolve the reducer body when it is local to this file;
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts:693]: ok

**Comment by rayhanadev** [comment]
bugbot run

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 10]
```diff
@@ -0,0 +1,711 @@
...
+                conditionState,
+              )
+            : [cloneReducerPathState(conditionState)];
+
+          nextStates.push(...consequentStates, ...alternateStates);
+          continue;
+        }
+
+        if (isNodeOfType(statement, "SwitchStatement")) {
+          // The switch value runs before any case.
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: why? either unnecesary comment or if there is a reason that switch needs to run before document why

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.test.ts [hunk 11]
```diff
@@ -0,0 +1,592 @@
+import { describe, expect, it } from "vite-plus/test";
+import { runRule } from "../../../test-utils/run-rule.js";
+import { noMutatingReducerState } from "./no-mutating-reducer-state.js";
+
+const run = (code: string) => runRule(noMutatingReducerState, code);
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.test.ts]: nit: unencesary abstraction

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+  //
+  // where `alias` was established by `const alias = state`.
+  if (isExpressionOriginalReducerStateReference(unwrappedNode, state)) return true;
+
+  if (isNodeOfType(unwrappedNode, "CallExpression")) {
+    // Object.assign returns its first argument, so this is still a same-reference
+    // return when the first argument is the original reducer state:
+    //
+    //   return Object.assign(state, patch);
+    if (isStaticMethodCallOnNamedObject(unwrappedNode, "Object", OBJECT_ASSIGN_METHOD)) {
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: idk if that assign method needs to be a const

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 13]
```diff
@@ -0,0 +1,711 @@
...
+  }
+
+  // Conditional/logical expressions may return the old state on just one side:
+  //
+  //   return changed ? { ...state } : state;
+  //   return maybeNext || state;
+  //
+  // If any possible branch returns the original reference, a prior mutation on
+  // this path is enough to report.
+  if (isNodeOfType(unwrappedNode, "ConditionalExpression")) {
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts:314]: i wonder if it makes sense to just do an exhaustive node check since the scope of this is wildly complex. bascially visit any amount of nodes in the scope and detect issue patterns. like this could probably be smarter. check the `oxc` rules for reference

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 14]
```diff
@@ -0,0 +1,711 @@
...
+          );
+          const mutationsAtReturn = [...activeState.mutations, ...returnMutations];
+          if (canExpressionReturnOriginalReducerStateReference(statement.argument, activeState)) {
+            reportReducerStateMutations(mutationsAtReturn);
+          }
+          continue;
+        }
+
+        if (isNodeOfType(statement, "IfStatement")) {
+          // The condition runs before either branch.
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: no

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts (hunk 13 above)
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts:314]: oh nvm you kind of do this already

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts [hunk 15]
```diff
@@ -0,0 +1,711 @@
...
+                blockState,
+                activeState,
+                blockScopedBindingNames,
+              ),
+            );
+          }
+          continue;
+        }
+
+        // Ordinary statements may mutate state and may also introduce aliases.
```
Comment (aidenybai) [inline packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-mutating-reducer-state.ts]: no
